src/solvers/fdsbp_unstructured/containers_2d.jl

```diff
@@ -0,0 +1,124 @@
+    calc_node_coordinates!(elements.node_coordinates, element, get_nodes(basis),
+                           corners_or_surface_curves)
+
+    calc_metric_terms!(elements.jacobian_matrix, element, basis,
+                       elements.node_coordinates)
+
+    calc_inverse_jacobian!(elements.inverse_jacobian, element, elements.jacobian_matrix)
+
+    calc_contravariant_vectors!(elements.contravariant_vectors, element,
+                                elements.jacobian_matrix)
+
+    calc_normal_directions!(elements.normal_directions, element,
+                            elements.jacobian_matrix)
+
+    return elements
+end
+
+# construct the metric terms for a FDSBP element "block". Directly use the derivative matrix
+# applied to the node coordinates.
+# TODO: FD; How to make this work for the upwind solver because basis has three available derivative matrices
+function calc_metric_terms!(jacobian_matrix, element, D_SBP::AbstractDerivativeOperator,
+                            node_coordinates)
+
+    # storage format:
+    #   jacobian_matrix[1,1,:,:,:] <- X_xi
+    #   jacobian_matrix[1,2,:,:,:] <- X_eta
+    #   jacobian_matrix[2,1,:,:,:] <- Y_xi
+    #   jacobian_matrix[2,2,:,:,:] <- Y_eta
+
+    # Compute the xi derivatives by applying D on the left
+    # This is basically the same as
+    # jacobian_matrix[1, 1, :, :, element] = Matrix(D_SBP) * node_coordinates[1, :, :, element]
+    # but uses only matrix-vector products instead of a matrix-matrix product.
+    for j in eachnode(D_SBP)
+        mul!(view(jacobian_matrix, 1, 1, :, j, element), D_SBP,
+             view(node_coordinates, 1, :, j, element))
+    end
+    # jacobian_matrix[2, 1, :, :, element] = Matrix(D_SBP) * node_coordinates[2, :, :, element]
+    for j in eachnode(D_SBP)
+        mul!(view(jacobian_matrix, 2, 1, :, j, element), D_SBP,
+             view(node_coordinates, 2, :, j, element))
+    end
+
+    # Compute the eta derivatives by applying transpose of D on the right
+    # jacobian_matrix[1, 2, :, :, element] = node_coordinates[1, :, :, element] * Matrix(D_SBP)'
+    for i in eachnode(D_SBP)
+        mul!(view(jacobian_matrix, 1, 2, i, :, element), D_SBP,
+             view(node_coordinates, 1, i, :, element))
+    end
+    # jacobian_matrix[2, 2, :, :, element] = node_coordinates[2, :, :, element] * Matrix(D_SBP)'
+    for i in eachnode(D_SBP)
+        mul!(view(jacobian_matrix, 2, 2, i, :, element), D_SBP,
+             view(node_coordinates, 2, i, :, element))
+    end
+
+    return jacobian_matrix
+end
+
+# construct the normal direction vectors (but not actually normalized) for a curved sided FDSBP element "block"
+# normalization occurs on the fly during the surface flux computation
+# OBS! This assumes that the boundary points are included.
+function calc_normal_directions!(normal_directions, element, jacobian_matrix)
+
+    # normal directions on the boundary for the left (local side 4) and right (local side 2)
+    N_ = size(jacobian_matrix, 3)
```

Comment:
Why `N_` instead of `N`?
It's also a bit confusing for me that you use `size(..., 3)` here but iterate over the fourth dimension. Maybe
```suggestion
    N = size(jacobian_matrix, 4)
```
?

Reply:
Yes, that is true. It would be clearer if this iterates over the fourth dimension (and the third dimension in the other loop). In the end the `N` value is always the same because we assume that the same number of nodes is used in each direction, but I will change this for the sake of clarity. I will also change all the `N_` variable names to `N`.